apps/srt-tunnel.cpp

```diff
@@ -0,0 +1,1074 @@
+    {
+        Verb() << "Closing SRT socket for " << uri();
+        lock_guard<mutex> lk(access);
+        if (m_socket == SRT_ERROR)
+            return;
+        srt_close(m_socket);
+        m_socket = SRT_ERROR;
+    }
+
+    virtual const char* type() override { return "srt"; }
+    virtual int ReadInternal(char* output, int size) override;
+    virtual bool IsErrorAgain() override;
+
+    virtual void Write(ref_t<bytevector> portion) override;
+    virtual void CreateListener() override;
+    virtual void CreateCaller() override;
+    virtual unique_ptr<Medium> Accept() override;
+    virtual void Connect() override;
+
+protected:
+    virtual void Init() override;
+
+    void ConfigurePre(SRTSOCKET socket);
+    void ConfigurePost(SRTSOCKET socket);
+
+    using Medium::Error;
+
+    static void Error(UDT::ERRORINFO& ri, const string& text)
+    {
+        throw TransmissionError("ERROR: " + text + ": " + ri.getErrorMessage());
+    }
+
+    virtual ~SrtMedium() override
+    {
+        Close();
+    }
+};
+
+class TcpMedium: public Medium
+{
+    int m_socket = -1;
+    friend class Medium;
+public:
+
+#ifndef _MSC_VER
+    using Medium::Medium;
+
+#else // MSVC not exactly supports C++11
+
+    TcpMedium(UriParser u, size_t ch): Medium(u, ch) {}
+
+#endif
+
+    bool IsOpen() override { return m_open; }
+    bool End() override { return m_eof; }
+    bool Broken() override { return m_broken; }
+
+    void Close() override
+    {
+        Verb() << "Closing TCP socket for " << uri();
+        lock_guard<mutex> lk(access);
+        if (m_socket == -1)
+            return;
+        ::close(m_socket);
+        m_socket = -1;
+    }
+
+    virtual const char* type() override { return "tcp"; }
+    virtual int ReadInternal(char* output, int size) override;
+    virtual bool IsErrorAgain() override;
+    virtual void Write(ref_t<bytevector> portion) override;
+    virtual void CreateListener() override;
+    virtual void CreateCaller() override;
+    virtual unique_ptr<Medium> Accept() override;
+    virtual void Connect() override;
+
+protected:
+
+    // Just models. No options are predicted for now.
+    void ConfigurePre(int )
+    {
+    }
+
+    void ConfigurePost(int)
+    {
+    }
+
+    using Medium::Error;
+
+    static void Error(int verrno, const string& text)
+    {
+        char rbuf[1024];
+        throw TransmissionError("ERROR: " + text + ": " + SysStrError(verrno, rbuf, 1024));
+    }
+
+    virtual ~TcpMedium()
+    {
+        Close();
+    }
+};
+
+void SrtMedium::Init()
+{
+    // This function is required due to extra option
+    // check need
+
+    if (m_options.count("mode"))
+        Error("No option 'mode' is required, it defaults to position of the argument");
+
+    if (m_options.count("blocking"))
+        Error("Blocking is not configurable here.");
+
+    // XXX
+    // Look also for other options that should not be here.
+
+    // Enforce the transtype = file
+    m_options["transtype"] = "file";
```

Comment:
`m_options["mode"] = caller` will do the same, as passing host "--" in `ConfigurePre(...)` bellow.